File: android/app/src/test/java/com/hawksnest/core/logic/DedupeTest.kt
package com.hawksnest.core.logic

import com.hawksnest.core.ha.HassEntity
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Test

class DedupeTest {

    private fun light(id: String, name: String) = HassEntity(
        entityId = id,
        state = "off",
        attributes = JsonObject(mapOf("friendly_name" to JsonPrimitive(name))),
    )

    @Test
    fun `a ring twin of an mqtt light is dropped — the mqtt one survives`() {
        val entities = mapOf(
            "light.front_light" to light("light.front_light", "Front Light"),
            "light.front_light_2" to light("light.front_light_2", "Front Light"),
        )
        val platforms = mapOf(
            "light.front_light" to MQTT_PLATFORM,
            "light.front_light_2" to RING_PLATFORM,
        )
        val out = dedupeRingMqtt(entities, platforms)
        assertEquals(setOf("light.front_light"), out.keys)
    }

    @Test
    fun `a ring entity with no mqtt twin survives`() {
        val entities = mapOf("light.porch" to light("light.porch", "Porch"))
        val out = dedupeRingMqtt(entities, mapOf("light.porch" to RING_PLATFORM))
        assertEquals(setOf("light.porch"), out.keys)
    }

    @Test
    fun `different names or domains never collide`() {
        val entities = mapOf(
            "light.front" to light("light.front", "Front Light"),
            "light.back" to light("light.back", "Back Light"),
            "switch.front_light" to HassEntity(
                entityId = "switch.front_light",
                state = "off",
                attributes = JsonObject(mapOf("friendly_name" to JsonPrimitive("Front Light"))),
            ),
        )
        val platforms = mapOf(
            "light.front" to MQTT_PLATFORM,
            "light.back" to RING_PLATFORM,
            "switch.front_light" to RING_PLATFORM,
        )
        assertEquals(entities.keys, dedupeRingMqtt(entities, platforms).keys)
    }

    @Test
    fun `non-ring platforms are untouched even on a name collision`() {
        val entities = mapOf(
            "light.hall" to light("light.hall", "Hall"),
            "light.hall_z" to light("light.hall_z", "Hall"),
        )
        val platforms = mapOf(
            "light.hall" to MQTT_PLATFORM,
            "light.hall_z" to "zwave_js",
        )
        assertEquals(entities.keys, dedupeRingMqtt(entities, platforms).keys)
    }

    @Test
    fun `empty platform map (registry not loaded yet) passes everything through`() {
        val entities = mapOf("light.front" to light("light.front", "Front Light"))
        assertEquals(entities.keys, dedupeRingMqtt(entities, emptyMap()).keys)
    }
}

File: src/lib/__tests__/dedupe.test.ts
import { describe, it, expect } from "vitest";
import { dedupeRingMqtt, RING_PLATFORM, MQTT_PLATFORM } from "../dedupe";
import type { HassEntity } from "../ha";

const light = (id: string, name: string): HassEntity => ({
  entity_id: id,
  state: "off",
  attributes: { friendly_name: name },
});

describe("dedupeRingMqtt", () => {
  it("drops the ring twin of an mqtt light — the mqtt one survives", () => {
    const entities = {
      "light.front_light": light("light.front_light", "Front Light"),
      "light.front_light_2": light("light.front_light_2", "Front Light"),
    };
    const platforms = {
      "light.front_light": MQTT_PLATFORM,
      "light.front_light_2": RING_PLATFORM,
    };
    expect(Object.keys(dedupeRingMqtt(entities, platforms))).toEqual([
      "light.front_light",
    ]);
  });

  it("keeps a ring entity with no mqtt twin", () => {
    const entities = { "light.porch": light("light.porch", "Porch") };
    expect(
      Object.keys(dedupeRingMqtt(entities, { "light.porch": RING_PLATFORM })),
    ).toEqual(["light.porch"]);
  });

  it("never collides across different names, domains, or non-ring platforms", () => {
    const entities = {
      "light.front": light("light.front", "Front Light"),
      "light.back": light("light.back", "Back Light"),
      "light.hall": light("light.hall", "Hall"),
      "light.hall_z": light("light.hall_z", "Hall"),
    };
    const platforms = {
      "light.front": MQTT_PLATFORM,
      "light.back": RING_PLATFORM,
      "light.hall": MQTT_PLATFORM,
      "light.hall_z": "zwave_js",
    };
    expect(Object.keys(dedupeRingMqtt(entities, platforms)).sort()).toEqual(
      Object.keys(entities).sort(),
    );
  });

  it("passes everything through before the registry loads (empty platforms)", () => {
    const entities = { "light.front": light("light.front", "Front Light") };
    expect(dedupeRingMqtt(entities, {})).toEqual(entities);
  });
});
